Added quick build feature

sinf opened the pull request:
Closes #7964.

Pressing the production tab hotkey selects the completed building so it can be placed.

sinf committed:
Added quick build feature

Mailaender commented:
:+1: Works as promised.

pchote commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -189,11 +210,9 @@ bool HandleEvent(ProductionIcon icon, bool isLeftClick, bool handleMultiple)
 				{
-					Sound.Play(TabClick);
-					World.OrderGenerator = new PlaceBuildingOrderGenerator(CurrentQueue, icon.Name);
+					// Pick up a completed building
```
Could you please adjust this so that it doesn't have an empty body?

sinf commented:
I reorganized HandleEvent's conditionals on pchote's request. Hopefully using bitwise ops isn't against our ~~religion~~ coding standards.  At least I got rid of a duplicate Sound.Play :)

Phrohdoh commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -228,12 +223,43 @@ bool HandleEvent(ProductionIcon icon, bool isLeftClick, bool handleMultiple)
+						break;
+					}
+
+					// falltrough
```
`// Fall through`, but is this comment even necessary?

sinf replied:
It is mostly a matter of opinion. Actually all of the comments seem somewhat redundant, no?

Phrohdoh replied:
Well I think this one is particularly redundant, but the others (such as `Left click and first != null`) may be more useful.

penev92 commented:
OK, this is a very nice feature and it seems to work ingame.
However, things like `var thing` and `var isLeft = Convert.ToInt32(isLeftClick);` make me feel uncomfortable with the implementation.

penev92 commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -182,41 +182,36 @@ public override bool HandleMouseInput(MouseInput mi)
+
+		public bool PickUpCompletedBuilding()
+		{
+			var thing = icons.Where(i => i.Value.Queued.Count > 0).SingleOrDefault(i => i.Value.Queued.FirstOrDefault().Done);
```
`.FirstOrDefault()` here is a potential NRE.

penev92 commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -182,41 +182,36 @@ public override bool HandleMouseInput(MouseInput mi)
+		public bool PickUpCompletedBuilding()
+		{
+			var thing = icons.Where(i => i.Value.Queued.Count > 0).SingleOrDefault(i => i.Value.Queued.FirstOrDefault().Done);
+			return thing.Value != null ? PickUpCompletedBuildingIcon(thing.Value) : false;
```
This can be simplified to `return thing.Value != null && PickUpCompletedBuildingIcon(thing.Value);`

sinf commented:
Changed HandleEvent again. Split into 2 functions for each mouse button, replaced the switch with ifs and replaced LINQ queries with a loop. Still seems to work as expected.

Phrohdoh commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -182,60 +182,93 @@ public override bool HandleMouseInput(MouseInput mi)
+		public bool PickUpCompletedBuilding()
+		{
+			var myIcons = icons.Values.ToArray();
+			for (var i = 0; i < icons.Count; i++)
```
This can be a `foreach`.

pchote commented:
> replaced LINQ queries with a loop

Why?

The other changes look like a nice improvement in code structure, but I haven't looked at them in detail.

pchote commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -182,60 +182,93 @@ public override bool HandleMouseInput(MouseInput mi)
+
+		public bool PickUpCompletedBuilding()
+		{
+			var myIcons = icons.Values.ToArray();
```
This forces the whole icons collection to be walked and copied to a new object, while using LINQ will only enumerate as far as it needs to find the completed building.

sinf commented:
I just discovered the foreach keyword and avoided the catastrophic performance loss with it.

penev92 commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -182,60 +182,91 @@ public override bool HandleMouseInput(MouseInput mi)
+
+			Sound.Play(TabClick);
+
+			// instant cancel of things we havent started yet and things that are finished
```
"Instant", "haven't", and also things that are paused. Preferably they should be ordered in che comment like they are in the code.

penev92 commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -182,60 +182,91 @@ public override bool HandleMouseInput(MouseInput mi)
+			return false;
+		}
+
+		public bool PickUpCompletedBuilding()
```
Does this method need to be bool?
I don't see the return values being used anywhere.

penev92 commented:
Code looks much better now, thanks.
I will test it again later, so far it has a :+0.5: from me.

sinf commented:
I fixed what penev didn't like. Also noticed that clicking a tab icon picked up buildings for placement so I fixed that too. It only makes sense to pick up buildings with either hotkeys or by clicking the production icon.

Phrohdoh commented:
I was under the impression that that was the whole point of this PR.

pchote commented:
>  Also noticed that clicking a tab icon picked up buildings for placement so I fixed that too.

This is intentional behavior, based on what the newer RTS games do.

pchote commented on OpenRA.Mods.Cnc/Widgets/Logic/ProductionTabsLogic.cs:
```diff
@@ -24,7 +24,7 @@ void SetupProductionGroupButton(ProductionTypeButtonWidget button)
 				return;
 
-			Action<bool> selectTab = reverse =>
+			Action<bool> selectTab = (reverse) =>
```
OpenRA code style is to omit parents when there is a single parameter.

pchote commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -182,60 +182,89 @@ public override bool HandleMouseInput(MouseInput mi)
-				}
-				else
-					Sound.Play(DisabledTabClick);
+				Sound.Play(TabClick);
```
This method is called from a bunch of places, so you shouldn't be doing high-level UI things here (playing the click sound).

sinf replied:
Refactoring the sound system is not in the scope of this PR.

penev92 replied:
It seems the old code does that, so I'm OK with this staying.

pchote replied:
I just want you to move `Sound.Play(TabClick);` outside of `PickUpCompletedBuildingIcon`.

sinf commented:
I changed the behaviour of mouse clicks back so buildings are picked when changing tabs by clicking tab icons.

sinf commented:
Removed Sound.Play from PickUpCompletedBuildingIcon. HandleEvent is the only one that plays sounds. According to my brief auditory inspection it matches the current openra behaviour.

penev92 commented:
Looks good to me now :+1:

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/d868aaf9f55b0ae848b122c3926c12224977049b...94de7d81f857ddbba750b6d107cf575d810e0870)
